security/rbac/engine/engine.go

```diff
@@ -0,0 +1,192 @@
+//  policy names that can't be evaluated due to missing attributes will be returned.
+// Else, there is not a match.
+func (engine rbacEngine) matches(evalMap map[string]interface{}) (bool, []string) {
+	matchingPolicyNames := []string{}
+	unknownPolicyNames := []string{}
+	for policyName, condition := range engine.conditions {
+		match, err := matchesCondition(condition, evalMap)
+		if err != nil {
+			unknownPolicyNames = append(unknownPolicyNames, policyName)
+		} else if match {
+			matchingPolicyNames = append(matchingPolicyNames, policyName)
+		}
+	}
+	if len(matchingPolicyNames) > 0 {
+		return true, matchingPolicyNames
+	}
+	return false, unknownPolicyNames
+}
+
+// CelEvaluationEngine is the struct for CEL engine.
+type CelEvaluationEngine struct {
+	engines []rbacEngine
+}
+
+// NewCelEvaluationEngine builds a cel evaluation engine from a list of Envoy RBACs.
+func NewCelEvaluationEngine(rbacs []pb.RBAC) (CelEvaluationEngine, error) {
+	if len(rbacs) < 1 || len(rbacs) > 2 {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "must provide 1 or 2 RBACs")
+	}
+	if len(rbacs) == 2 && (rbacs[0].Action != pb.RBAC_DENY || rbacs[1].Action != pb.RBAC_ALLOW) {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "when providing 2 RBACs, must have 1 DENY and 1 ALLOW in that order")
+	}
+	var engines []rbacEngine
+	for _, rbac := range rbacs {
+		engines = append(engines, newRbacEngine(rbac))
+	}
+	return CelEvaluationEngine{engines}, nil
+}
+
+// Evaluate is the core function that evaluates whether an RPC is authorized.
+//
+// ALLOW policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns allow. If all of the RBAC conditions are
+// evaluated as false, then it returns deny. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns deny. If all of the RBAC conditions are
+// evaluated as false, then it returns allow. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy + ALLOW policy. Evaluation is in the following order: If one
+// of the expressions in the DENY policy is true, the authorization engine
+// returns deny. If one of the expressions in the DENY policy is unknown, it
+// returns undecided. Now all the expressions in the DENY policy are false,
+// it returns the evaluation of the ALLOW policy.
+func (celEngine CelEvaluationEngine) Evaluate(args AuthorizationArgs) (AuthorizationDecision, error) {
+	evalMap := args.toEvalMap()
+	numEngines := len(celEngine.engines)
+	if numEngines < 1 || numEngines > 2 {
+		return AuthorizationDecision{}, status.Errorf(codes.Internal, "each CEL engine should have 1 or 2 RBAC engines; instead, there are %d in the CEL engine provided", numEngines)
+	}
+	for _, engine := range celEngine.engines {
```

Comment:
This can be simplified as follows:
evaluate the first celEngine. 
If Decision is Deny or Unknown, return directly.
If first celEngine evaluated as Allow and there is a second celEngine, return Decision of second celEngine.